File: ch3/akkademaid-java/src/test/java/com/akkademy/askdemo/AskDemoArticleParser.java
package com.akkademy.askdemo;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.ActorSelection;
import akka.japi.pf.ReceiveBuilder;
import akka.util.Timeout;
import scala.PartialFunction;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

import static scala.compat.compat.FutureConverters.*;
import static akka.pattern.Patterns.ask;

public class AskDemoArticleParser extends AbstractActor {

    private final ActorSelection cacheActor;
    private final ActorSelection httpClientActor;
    private final ActorSelection artcileParseActor;
    private final Timeout timeout;

    public AskDemoArticleParser(String cacheActorPath, String httpClientActorPath, String artcileParseActorPath, Timeout timeout) {
        this.cacheActor = context().actorSelection(cacheActorPath);
        this.httpClientActor = context().actorSelection(httpClientActorPath);
        this.artcileParseActor = context().actorSelection(artcileParseActorPath);
        this.timeout = timeout;
    }

    public PartialFunction receive() {
        return ReceiveBuilder.
                match(ParseArticle.class, msg -> {

                    final CompletionStage cacheResult = toJava(ask(cacheActor, msg.url, timeout));
                    final CompletionStage result = cacheResult.handle((x, t) -> {
                        return (x != null)
                                ? CompletableFuture.completedFuture(x)
                                : toJava(ask(httpClientActor, msg.url, timeout)).
                                thenCompose(rawArticle -> toJava(ask(artcileParseActor, (String) rawArticle, timeout)))
                                ;
                    }).thenCompose(x -> x);

                    final ActorRef senderRef = sender();
                    result.handle((x,t) -> {
                        if(x == null)
                            senderRef.tell(x, self());
                        else
                            senderRef.tell(new akka.actor.Status.Failure(t), self());
                        return null;
                    });

                }).build();
    }
}

File: ch3/akkademaid-scala/src/main/scala/com/akkademy/askdemo/AskDemoArticleParser.scala
package com.akkademy.askdemo

import akka.actor.Actor
import akka.pattern.ask
import akka.util.Timeout
import com.akkademy.sapi.{SetRequest, GetRequest}

class AskDemoArticleParser(cacheActorPath: String,
                           httpClientActorPath: String,
                           acticleParserActorPath: String,
                           implicit val timeout: Timeout
                            ) extends Actor {
  val cacheActor = context.actorSelection(cacheActorPath)
  val httpClientActor = context.actorSelection(cacheActorPath)
  val articleParserActor = context.actorSelection(cacheActorPath)
  import scala.concurrent.ExecutionContext.Implicits.global

  override def receive: Receive = {
    case ParseArticle(uri) =>
      val senderRef = sender() //sender ref needed!!

      val cacheResult = cacheActor ? GetRequest(uri)

      val result = cacheResult.recoverWith {
        case _: Exception =>
          val fRawResult = httpClientActor ? uri
          fRawResult flatMap (rawArticle =>
            articleParserActor ? rawArticle)
      }

      result.mapTo[String] onComplete {
        case scala.util.Success(x) =>
          cacheActor ! SetRequest(uri, x)
          senderRef ! x
        case scala.util.Failure(t) =>
          senderRef ! akka.actor.Status.Failure(t)
      }
  }
}
